mpc-core/src/garble/label.rs

```diff
@@ -0,0 +1,451 @@
+        Self {
+            input,
+            labels: labels.to_vec(),
+        }
+    }
+
+    pub fn id(&self) -> usize {
+        self.input.id
+    }
+}
+
+impl InputLabels<WireLabelPair> {
+    /// Generates a full set of input [`WireLabelPair`] for the provided [`Circuit`]
+    pub fn generate<R: Rng + CryptoRng>(
+        rng: &mut R,
+        circ: &Circuit,
+        delta: Option<Delta>,
+    ) -> (Vec<Self>, Delta) {
+        let (labels, delta) = WireLabelPair::generate(rng, delta, circ.input_len(), 0);
+
+        // This should never panic due to invariants enforced during the construction of a `Circuit`
+        let inputs: Vec<InputLabels<WireLabelPair>> = circ
+            .inputs()
+            .iter()
+            .map(|input| InputLabels::new(input.clone(), &pick(&labels, input.as_ref().wires())))
+            .collect();
+
+        (inputs, delta)
+    }
+
+    /// Generates a full set of input [`WireLabelPair`] for the provided [`Circuit`], split by provided input ids.
+    /// The first collection corresponds to the provided ids, the other collection is the remainder.
+    pub fn generate_split<R: Rng + CryptoRng>(
+        rng: &mut R,
+        circ: &Circuit,
+        input_ids: &[usize],
+        delta: Option<Delta>,
+    ) -> Result<((Vec<Self>, Vec<Self>), Delta), Error> {
+        let mut input_ids = input_ids.to_vec();
+        input_ids.sort();
+        input_ids.dedup();
+        for id in input_ids.iter() {
+            if circ.input(*id).is_none() {
+                return Err(Error::InvalidInput(InputError::InvalidId(*id)));
+            }
+        }
+
+        let (labels, delta) = Self::generate(rng, circ, delta);
+
+        let (left, right): (Vec<Self>, Vec<Self>) = labels
+            .into_iter()
+            .partition(|labels| input_ids.contains(&labels.id()));
+
+        Ok(((left, right), delta))
+    }
+
+    /// Returns input wire labels corresponding to an [`InputValue`]
+    pub fn select(&self, value: &InputValue) -> Result<InputLabels<WireLabel>, Error> {
+        // TODO: Don't panic, return proper error
+        assert_eq!(value.id(), self.input.id);
+
+        let labels: Vec<WireLabel> = self
+            .labels
+            .iter()
+            .zip(value.as_ref().iter())
+            .map(|(pair, value)| pair.select(*value))
+            .collect();
+
+        Ok(InputLabels {
+            input: self.input.clone(),
+            labels,
+        })
+    }
+}
+
+impl<T> AsRef<[T]> for InputLabels<T> {
+    fn as_ref(&self) -> &[T] {
+        &self.labels
+    }
+}
+
+/// Input labels that have been sanitized are safe to use to evaluate a garbled circuit
+///
+/// It is important to check that the generator has provided the expected input labels,
+/// otherwise they may have an opportunity to behave maliciously to extract the evaluator's
+/// private inputs.
+#[derive(Debug, Clone)]
+pub struct SanitizedInputLabels(Vec<WireLabel>);
```

Comment:
Could we please have tests which trigger all the error conditions, since this code is security-critical.